Buildinfo since endpoint

Foxboron opened the pull request:
Implements a endpoint returning a list of buildinfo files since a given
date given in Unix time.

Example: GET /api/buildinfo/since/0

Signed-off-by: Morten Linderud <morten@linderud.pw>

Foxboron committed:
First revision of buildinfo since endpoint

Implements a endpoint returning a list of buildinfo files since a given
date given in Unix time.

Example: GET /api/buildinfo/since/0

Signed-off-by: Morten Linderud <morten@linderud.pw>

kpcyrd commented:
Random thought, but it might be a good idea to:
- limit the number of buildinfo files returned (eg 100, oldest first)
- return the timestamps of the database entries as well

That would allow me to start at a low number without overloading the database and then slowly walk through all pages until an empty list is returned, then poll every hour until new entries are returned.

lamby commented:
What is the use-case here? You want to know about new buildinfo files? If so, there are surely better ways of doing this than polling.

kpcyrd commented:
This is needed for #48, which was filed for https://salsa.debian.org/reproducible-builds/debian-rebuilder-setup/tree/integrate-srebuild

bidb pushing the info to us, the rebuilder, would work as well but that would require coordination with you to run a rebuilder. Ideally, anybody who wants to run a rebuilder should be able to get that data without having to coordinate anything.

Something like server-sent-events or websockets would technically avoid polling as well, but require more work from both sides. :)

lamby commented:
(As an aside, can I request `bidb` → `buildinfo.debian.net`. Latter is the proper service name that others will understand but the former is "just" the Python module name.)

lamby commented:
(Dropped "first revision of" prefix)

SantiagoTorres committed:
api:buildinfo/since add 'created' field

Adds a created-since field to the api response, filters by gte (so as to
avoid microsecond time drift) and sorts the results.

SantiagoTorres committed:
api/since: add limit parameter

SantiagoTorres commented:
@Foxboron please take a look at https://github.com/Foxboron/buildinfo.debian.net/pull/1 to address soem of the remaining issues...

Foxboron committed:
Added missing values from model

Signed-off-by: Morten Linderud <morten@linderud.pw>

h01ger commented:
ping, Foxboron and kpcyrd would be very happy to get this merged :)

lamby commented on bidb/api/views.py:
```diff
@@ -18,3 +24,38 @@ def submit(request):
+    try:
+        created = datetime.datetime.fromtimestamp(int(date) + 1)
+    except ValueError:
+        raise SuspiciousOperation("Date value should be an integer")
```
I suggest returning `HTTPBadRequest` rather than `SuspiciousOperation` in this case:

```suggestion
        return HttpResponseBadRequest("Date value should be an integer")
```

lamby commented on bidb/api/urls.py:
```diff
@@ -5,4 +5,6 @@
 urlpatterns = (
     url(r'^api/submit$', views.submit,
         name='submit'),
+    url(r'^api/buildinfo/since/(?P<date>[^/]+)$', views.buildinfo_since,
```
Given that we know we want an integer here, I suggest:

```suggestion
    url(r'^api/buildinfo/since/(?P<date>\d+)$', views.buildinfo_since,
```

lamby commented on bidb/api/views.py:
```diff
@@ -18,3 +24,38 @@ def submit(request):
+    try:
+        limit = int(request.GET.get('limit', 100))
+        if limit <= 0:
+            raise SuspiciousOperation("Limit value should be positive")
```
Ditto:

```suggestion
            return HttpResponseBadRequest("Limit value should be positive")
```

lamby commented on bidb/api/views.py:
```diff
@@ -18,3 +24,38 @@ def submit(request):
+        raise SuspiciousOperation("Limit value should be an integer")
+
+    buildinfo = Buildinfo.objects.filter(
+            created__gte=created
```
Double indentation, and missing trailing comma:
```suggestion
            created__gte=created,
```

lamby commented on bidb/api/views.py:
```diff
@@ -18,3 +24,38 @@ def submit(request):
+        'source',
+        'architecture',
+    ).order_by('created')[:limit]
+    return JsonResponse({'buildinfos': [{
```
Please add a space before this.

lamby commented on bidb/api/views.py:
```diff
@@ -18,3 +24,38 @@ def submit(request):
+        'source': x.source.name,
+        'version': x.version,
+        'architecture': x.architecture.name,
+        'created':  time.mktime(x.created.timetuple())
```
Missing trailing comma 
```suggestion
        'created':  time.mktime(x.created.timetuple()),
```

SantiagoTorres commented:
Likewise, I could stop replaying information on buildinfo.nyu.wtf when
this happens.

Foxboron, kpcyrd, I can take a dive at these fixes tonight, would you
like to do it sooner?

Thanks,
-Santiago.

On Thu, Dec 13, 2018 at 02:42:57AM -0800, Holger Levsen wrote:
> ping, Foxboron and kpcyrd would be very happy to get this merged :)
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/lamby/buildinfo.debian.net/pull/54#issuecomment-446923133

Foxboron commented:
@lamby I have squashed the code review commit
